source/includes/inkling2-reference/_functions.html.md

```diff
@@ -0,0 +1,233 @@
+var ZeroTo100: number<0..100> = OneOrTWo
+
+# This will generate a compiler error because "ZeroTo100"
+# may not match the values of the enumerated
+# type of "OneOrTWo".
+OneOrTWo = ZeroTo100
+
+var ZeroTo50: number<0..50 step 1> = 50
+
+# This will generate a compiler warning because "ZeroTo50"
+# may require clipping to fit within the specified range.
+ZeroTo50 = ZeroTo100
+```
+
+For more details about compatible types, refer to the [types documentation][1].
+
+
+#### Runtime Errors & Debugging
+The Inkling compiler attempts to detect and report most errors statically. However, there are certain circumstances where runtime may occur. Examples include undefined mathematical operations like division by zero. Runtime errors are logged and reported during training and may terminate a training session.
+
+
+### Function Statements
+
+#### Variable Declarations
+
+Variables can be declared within a function. All variables have function-level scope even if they are declared within a sub-scope such as an if/else statement. Each variable within a function must have a unique name and a type. The type can be omitted if it can be inferred from an initial value assigned to the variable.
+
+```inkling2--syntax
+variableDeclaration :==
+  var <varName> [':' <varType>] ['=' <expression>]
+```
+
+A variable that is not assigned an initial value must be assigned a value along all code paths before it is used.
+
+```inkling2--code
+# Variable with no initial assignment.
+var TwoOrThree: number<2, 3>
+
+if IsCondition() {
+  TwoOrThree = 2
+}
+
+# This will generate a compiler error because the
+# variable may not have a defined value.
+return TwoOrThree
+```
+
+#### Variable Assignments
+A previously-declared variable can be assigned a new value as long as that value matches the type of the variable.
+
+```inkling2--syntax
+variableAssignment :==
+  <varName> '=' <expression>
+```
+
+#### Conditional Execution
+Inkling supports if/else statements, which can be chained together with “else if” clauses. The braces that surround the code blocks within an `if` or `else` clause are required.
+
+```inkling2--syntax
+conditionalStatement :==
+  if <expression> '{'
+    [functionStatement]*
+  '}' [else if <expression> '{'
+    [functionStatement]*
+  '}']* [else '{'
+    [functionStatement]*
+  '}']
+```   
+
+#### Return Statements
+A return statement terminates execution of the function and returns the specified value to its caller. The type of the expression must match the declared return type of the function.
+
+```inkling2--syntax
+returnStatement :==
+  return <expression>
+```
+
+### Expressions
+Inkling supports basic expressions and a variety of unary and binary operators.
+
+#### Built-in Operators
+| Operation        | Syntax        | Description |
+| -                | -             | -           |
+| Function call    | X()           | Invokes a function with zero or more arguments |
+| Array index      | X[Y]          | Indexes into an array |
```

Comment:
remove until we support arrays? or add `(future)`

Reply:
We fully support arrays within functions. They're not yet supported for types passed to/from the sim (because we need SDK3) or concept inputs/outputs (because the AI engine doesn't yet know how to encode them to/from tensors). The inkling compiler will provide an error in these specific cases if arrays or structs are used. But they're safe to use everywhere else.